MAD Quests

LegitDongo opened the pull request:
# Description
Quests!
Merry Christmas guys, this gives MAD compatibility for quests!

Check out the included documentation on the new DTS and filters. These should be pretty inclusive to what you guys want, if there's more that's needed, let me know and I'll get it added.

Translations are probably a little rough (even though most of them I got from the game), so please comment where they are wrong.

## Type of Change
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (would cause existing functionality to change)

## Motivation and Context
Because people want them quests

## How Has This Been Tested?
Very lightly, windows 10, webhook tester

## Wiki Update
- [x] This change requires an update to the Wiki.
- [ ] This change does not require an update to the Wiki.

Wiki update already done

LegitDongo committed:
Quests
Quest documentation

LegitDongo committed:
Merge branch 'dev' of https://github.com/PokeAlarm/PokeAlarm into mad_quests

kvangent commented on PokeAlarm/Events/QuestEvent.py:
```diff
@@ -19,10 +21,9 @@ def __init__(self, data):
+            str, data.get('pokestop_name', data.get('name')), Unknown.REGULAR)
         self.stop_image = check_for_none(
-            str, data.get('pokestop_url') or data.get('url'), Unknown.REGULAR)
+            str, data.get('pokestop_url', data.get('url')), Unknown.REGULAR)
```
This looks wrong - what are you trying to do?

LegitDongo replied:
see #691 - sometimes `or` evaluates things as falsey, this seems like the best remedy

kvangent commented on PokeAlarm/Events/QuestEvent.py:
```diff
@@ -38,28 +39,43 @@ def __init__(self, data):
-                (datetime.combine(date.today(), time(23, 59)) -
-                 datetime(1970, 1, 1)).total_seconds()))
-        self.reward_type = check_for_none(int, data.get('type'), 0)
+        self.quest_type_raw = data['quest_type']
```
Are all these fields requires required? right now it'll throw an error for any that don't exist

kvangent replied:
These should probably be .gets for backwards compatibility

kvangent commented on PokeAlarm/Utilities/QuestUtils.py:
```diff
@@ -27,3 +27,56 @@ def get_reward_type(reward_type):
+
+
+def reward_string(quest, locale):
+    if quest.reward_type_id == 7:
```
Can you add some comments or clarification to this function? Not quite sure what the significance of these ids are, or why they mean different things.

kvangent commented on docs/configuration/alarms/discord.rst:
```diff
@@ -193,6 +193,15 @@ Example: Alarm Configuration Using Optional Parameters
+      "quests":{
+          "webhook_url":"DISCORD_WEBHOOK_URL_FOR_INVASION_CHANNEL",
+          "username":"Quest",
+          "icon_url*":"<YOUR CUSTOM URL HERE>/<quest_image>.png",
```
We should be consistent with values we want the user to replace - also consider avoiding `<>` as that currently signals a DTS. 
```suggestion
          "icon_url*":"<YOUR_CUSTOM_URL_HERE>/<quest_image>.png",
```

LegitDongo committed:
Merge branch 'dev' into mad_quests

LegitDongo commented:
For those having issues applying this patch, try this:
`git apply --reject --whitespace=fix 706.patch`
this will fix whitespace issues and make new files for rejected changes

en.json will need things replaced based off rejected changes
`quest_types` -> `quest_reward_types`

LegitDongo committed:
Add documentation to go with scattered IDs
Change a lot of the fields to be optional - everything still works right
Change documentation across the board to not make `<YOUR_CUSTOM_URL_HERE>` not DTS-like

LegitDongo committed:
Merge branch 'dev' into mad_quests

LegitDongo committed:
Re-add costumes with docs
Small cleanup

LegitDongo committed:
Fix lng_5

LegitDongo committed:
Merge branch 'dev' into mad_quests

LegitDongo committed:
Merge branch 'dev' into mad_quests

LegitDongo committed:
Add task_contains and task_excludes for basic task checking

LegitDongo committed:
Merge branch 'dev' into mad_quests

LegitDongo committed:
Change default alarm from `quest_condition` to `quest_task`
Made everything else optional

LegitDongo committed:
Merge remote-tracking branch 'origin/mad_quests' into mad_quests_fix

LegitDongo committed:
Merge branch 'dev' into mad_quests
